parlai/chat_service/core/chat_service_manager.py

```diff
@@ -0,0 +1,677 @@
+
+    @abstractmethod
+    def _handle_bot_read(self, agent_id):
+        """Use this function to handle/execute events once the bot has observed the message."""
+
+    @abstractmethod
+    def _confirm_message_delivery(self, event):
+        """A callback for when messages are marked as delivered"""
+
+    def _handle_message_read(self, event):
+        # If the message was sent by another user (as in during a conversation)
+        # then we need to propogate the read back to that user.
+        reader = event['sender']['id']
+        agent_state = self.get_agent_state(reader)
+        if agent_state is None:
+            return
+        agent = agent_state.get_active_agent()
+        if agent is not None:
+            for partner in agent.message_partners:
+                # We don't know who sent the message that was seen, but we can
+                # send a message observed event to everyone else in the chat
+                self.message_sender.send_read(partner.id)
+
+    def _on_first_message(self, message):
+        """Handle first message from player.
+
+        Run when a psid is given that is not paired with any assignment yet.
+        Launch an overworld, complete onboarding, etc.
+
+        :param message:
+            message sent from agent
+        """
+        if self.service_reference_id is None:
+            self.service_reference_id = message['recipient']['id']
+
+        agent_id = message['sender']['id']
+        if self.opt['password'] is not None:
+            if message['message']['text'] != self.opt['password']:
+                self.observe_message(
+                    agent_id,
+                    'Sorry, this conversation bot is password-protected. If '
+                    'you have the password, please send it now.',
+                )
+                return
+
+        task_id = 'overworld-{}-{}'.format(agent_id, time.time())
+        agent = self._create_agent(task_id, agent_id)
+        agent_state = AgentState(agent_id, agent)
+        if self.opt['password'] is not None:
+            agent_state.stored_data['first_message'] = message
+        self.messenger_agent_states[agent_id] = agent_state
+
+        # launch overworld
+        _ = self.world_runner.launch_overworld(
+            task_id, self.overworld, self.onboard_map, agent
+        )  # labeled as future but lint throws error
+
+    def _on_new_message(self, message):
+        """Put an incoming message onto the correct agent's message queue.
+
+        :param message:
+            message to put on queue
+        """
+        agent_id = message['sender']['id']
+        if agent_id not in self.messenger_agent_states:
+            self._on_first_message(message)
+            return
+
+        agent_state = self.get_agent_state(agent_id)
+        if agent_state.get_active_agent() is None:
+            # return agent to overworld
+            if (
+                'text' in message['message']
```

Comment:
ditto here on `message['message']`

Reply:
Good point does using a get fix this ?
eg:
```python 
message.get("text", "")
```

Reply:
this was more so referring to the `message` field